Handle array parameters

agyoungs opened the pull request:
@malban This should handle list parameters for various types. I didn't implement any special handling like min/max for Numeric types, but that wouldn't be hard to add. It may make sense to promote some of that logic outside of the templated NumericParameter class to avoid duplicating code.

agyoungs committed:
Handle array parameters

agyoungs committed:
'Use' variable to get rid of warning

agyoungs committed:
Update params to handle int and int64 as needed

agyoungs committed:
Rename system ints to just int (normal for c++) and specified 64 for the 64 bit int parameters

agyoungs committed:
Revert "Rename system ints to just int (normal for c++) and specified 64 for the 64 bit int parameters"

This reverts commit 904895949e8d41ab4c9a8ffb5041f4f190fb615d.

agyoungs commented:
This now handles both ints and int64 types. This is helpful to maintain backwards compatibility with nodes that expect `int` to work for parameters (ROS2 uses the `int64_t` for all integer parameters. Since array handling is new I didn't implement a way to handle `std::vector<int>` (only std::vector<int64_t> is supported) since there isn't any code to break in existing nodes. Support for this requires more work than supporting int since the compiler will support implicit casting from `int64_t` to `int`, but has no way to do this for `std::vector<int64_t>` to `std::vector<int>`. It's pretty trivial, but I'm not sure it's worth supporting. The one side effect is when using the signautre `param("int64_array_param", {0}, "an int64 array")` your compiler will likely give you an ambiguous function signature error. You'll have to add an `L` to the `0` to make sure it uses the int64_t signature instead of the one for bools. Standard compilation seems to work for `int` in this case, but not `int64_t`. I don't see this as a big issue, though. Future code will just have to specify long int literals for arrays

agyoungs committed:
Update documentation

agyoungs committed:
Update documentation

agyoungs committed:
Formatting

agyoungs committed:
Refactored logic to override for vector type rather than duplicated code through inheritance structure for arrays

agyoungs committed:
typo

agyoungs committed:
Cleaned up templating

malban commented:
@agyoungs If you ended up adding array range and step constraints, can you update the documentation to reflect that?

agyoungs committed:
Fixed toString; Added examples; Documented new feautres

agyoungs commented:
> @agyoungs If you ended up adding array range and step constraints, can you update the documentation to reflect that?

Updated, and doing so made me realize that one of the toString methods I modified wasn't being used properly, so I fixed that as well (mainly just affected enums)

agyoungs committed:
Update README.md

typo
